synapse/push/mailer.py

```diff
@@ -123,6 +124,73 @@ def __init__(self, hs, app_name, template_html, template_text):
 
         logger.info("Created Mailer for app_name %s" % app_name)
 
+    @defer.inlineCallbacks
+    def send_threepid_validation(
+        self, email, client_secret, send_attempt, next_link=None
+    ):
+        """Send a threepid validation email for password reset or
+        registration purposes
+
+        Args:
+            email (str): The user's email address
+            client_secret (str): The provided client secret
+            send_attempt (int): Which send attempt this is
+
+        Returns:
+            The new session_id upon success
+
+        Raises:
+            SynapseError is an error occurred when sending the email
+        """
+        # Check that this email/client_secret/send_attempt combo is new or
+        # greater than what we've seen previously
+        session = yield self.store.get_threepid_validation_session(
+            "email", client_secret, address=email, validated=False
+        )
+
+        # Check to see if a session already exists and that it is not yet
+        # marked as validated
+        if session and session.get("validated_at") is None:
+            session_id = session["session_id"]
+            last_send_attempt = session["last_send_attempt"]
+
+            # Check that the send_attempt is higher than previous attempts
+            if send_attempt <= last_send_attempt:
+                # If not, just return a success without sending an email
+                return session_id
+        else:
+            # An non-validated session does not exist yet.
+            # Generate a session id
+            session_id = random_string(16)
+
+        # Generate a new validation token
+        token = random_string(32)
+
+        # Send the mail with the link containing the token, client_secret
+        # and session_id
+        try:
+            yield self.send_password_reset_mail(email, token, client_secret, session_id)
```

Comment:
I'm a bit confused about why this function claims to be used for "registration purposes" but we are sending a password reset email?